internal/server/server_functions.go

```diff
@@ -0,0 +1,502 @@
+		return
+	}
+
+	// Add the cluster if it doesn't exist
+	exists := false
+	for _, name := range s.clusters {
+		if name == clusterName {
+			exists = true
+			break
+		}
+	}
+
+	if !exists {
+		s.clusters = append(s.clusters, clusterName)
+		logger.Log.Info("Added new cluster to list", zap.String("cluster", clusterName))
+		s.broadcastClusters()
+	} else {
+		logger.Log.Debug("Cluster already in list", zap.String("cluster", clusterName))
+	}
+}
+
+// Function to handle deleting a cluster
+func (s *Server) deleteCluster(obj interface{}) {
+	logger.Log.Debug("deleteCluster called with object")
+
+	if obj == nil {
+		logger.Log.Warn("Received nil object in deleteCluster")
+		return
+	}
+
+	var clusterName string
+
+	// Handle the DeletedFinalStateUnknown case first
+	if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+		logger.Log.Debug("Received tombstone object in deleteCluster")
+		if unstructuredObj, ok := tombstone.Obj.(*unstructured.Unstructured); ok {
+			clusterName = unstructuredObj.GetName()
+			logger.Log.Info("Extracted cluster name from tombstone", zap.String("cluster", clusterName))
+		} else {
+			logger.Log.Warn("Tombstone contains unknown object type")
+			return
+		}
+	} else if unstructuredObj, ok := obj.(*unstructured.Unstructured); ok {
+		clusterName = unstructuredObj.GetName()
+		logger.Log.Info("Processing cluster deletion", zap.String("cluster", clusterName))
+	} else {
+		logger.Log.Warn("Received unknown object type in deleteCluster")
+		return
+	}
+
+	if clusterName == "" {
+		logger.Log.Warn("Could not determine cluster name for deletion")
+		return
+	}
+
+	// Remove the cluster from the list if it exists
+	for i, name := range s.clusters {
+		if name == clusterName {
+			s.clusters = append(s.clusters[:i], s.clusters[i+1:]...)
+			logger.Log.Info("Removed cluster from list", zap.String("cluster", clusterName))
+
+			// Also remove from conditions map to prevent memory leaks
+			if _, exists := s.clusterConditions[clusterName]; exists {
+				delete(s.clusterConditions, clusterName)
+				logger.Log.Info("Removed cluster conditions for", zap.String("cluster", clusterName))
+			}
+
+			s.broadcastClusters()
+			break
+		}
+	}
+}
+
+// Helper function to broadcast the current list of clusters to all clients
+func (s *Server) broadcastClusters() {
+	s.clientsMutex.Lock()
+	defer s.clientsMutex.Unlock()
+
+	for client := range s.clients {
+		err := client.conn.WriteJSON(map[string]interface{}{
+			"type":     "clusters",
+			"clusters": s.clusters,
+		})
+		if err != nil {
+			logger.Log.Error("Error sending cluster update", zap.Error(err))
+			client.conn.Close()
+			delete(s.clients, client)
+		}
+	}
+}
+
+// Function to handle cluster conditions using the object from the informer
+func (s *Server) updateConditions(obj interface{}) {
+	logger.Log.Debug("updateConditions called with object")
+
+	if obj != nil {
```

Comment:
You can use early returning to clean up the heavily nested if statements. For example instead of :
```suggestion
	if obj != nil {
```

```
if obj != nil {
    <do stuff>
}
``` 

You can have:
```
if obj == nil {
    return 
}

<do stuff>
```